winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,348 @@
+}
+
+static int current_is_wow = -1;
+static int is_32_bit_os = -1;
+
+static BOOL get_wow(HANDLE process, BOOL& is_wow, int& bitness)
+{
+  if (is_32_bit_os == -1)
+    {
+      SYSTEM_INFO info;
+
+      GetNativeSystemInfo(&info);
+      if (info.wProcessorArchitecture == 0)
+	is_32_bit_os = 1;
+      else if (info.wProcessorArchitecture == 9)
+	is_32_bit_os = 0;
+      else
+	is_32_bit_os = -2;
+    }
+
+  if (current_is_wow == -1 &&
+      !IsWow64Process (GetCurrentProcess (), &current_is_wow))
+    current_is_wow = -2;
+
+  if (is_32_bit_os == -2 || current_is_wow == -2)
+    return FALSE;
+
+  if (!IsWow64Process (process, &is_wow))
+    return FALSE;
+
+  bitness = is_32_bit_os || is_wow ? 32 : 64;
+  return TRUE;
+}
+
+static LPTHREAD_START_ROUTINE
+get_exit_process_address_for_process(HANDLE process, int exit_code, DWORD pid)
+{
+  BOOL is_wow;
+  int bitness;
+  if (!get_wow (process, is_wow, bitness))
+    return NULL; /* could not determine */
+
+  if (is_wow == current_is_wow)
+    {
+      static LPTHREAD_START_ROUTINE exit_process_address;
+      if (!exit_process_address)
+	{
+	  HINSTANCE kernel32 = GetModuleHandle ("kernel32");
+	  exit_process_address = (LPTHREAD_START_ROUTINE)
+	    GetProcAddress (kernel32, "ExitProcess");
+	}
+      return exit_process_address;
+    }
+
+  /* Trying to terminate a 32-bit process from 64-bit or vice versa */
+  static LPTHREAD_START_ROUTINE exit_process_address;
+  if (!exit_process_address)
+    {
+      exit_process_address =
+	kill_via_console_helper (bitness, L"ExitProcess", exit_code, pid);
+    }
+  return exit_process_address;
+}
+
+static LPTHREAD_START_ROUTINE
+get_ctrl_routine_address_for_process_and_inject_remote_thread(HANDLE process, int exit_code, DWORD pid)
+{
+  BOOL is_wow;
+  int bitness;
+  if (!get_wow (process, is_wow, bitness))
+    return NULL; /* could not determine */
+
+  if (bitness == 64)
+    {
+      static LPTHREAD_START_ROUTINE ctrl_routine_address;
+      if (!ctrl_routine_address)
+	ctrl_routine_address =
+	  kill_via_console_helper (64, L"CtrlRoutine", exit_code, pid);
+      return ctrl_routine_address;
+    }
+
+  static LPTHREAD_START_ROUTINE ctrl_routine_address;
+  if (!ctrl_routine_address)
+    {
+      ctrl_routine_address =
+	kill_via_console_helper (32, L"CtrlRoutine", exit_code, pid);
+    }
+  return ctrl_routine_address;
+}
+
+
+/**
+ * Terminates the process corresponding to the process ID
+ *
+ * This way of terminating the processes is not gentle: the process gets
+ * no chance of cleaning up after itself (closing file handles, removing
+ * .lock files, terminating spawned processes (if any), etc).
+ */
+static int
+exit_one_process(HANDLE process, int exit_code, DWORD pid)
+{
+  LPTHREAD_START_ROUTINE address = NULL;
+  int signo = exit_code & 0x7f;
+
+  switch (signo)
+    {
+      case SIGINT:
+      case SIGQUIT:
+	address = get_ctrl_routine_address_for_process_and_inject_remote_thread(process, signo == SIGINT ?
```

Comment:
Since we do not use that address at all anymore, this call to `get_ctrl_routine_...()` could be replaced by a simple

```c
if (get_wow(process, is_wow, bitness) &&
    0 /* exit code 0 = success */ == kill_via_console_helper (bitness, L"CtrlRoutine",
                             signo == SIGINT ? CTRL_C_EVENT : CTRL_BREAK_EVENT, pid)))
            return 0;
```